http-client/src/main/java/io/airlift/http/client/StreamingResponseHandler.java

```diff
@@ -0,0 +1,101 @@
+package io.airlift.http.client;
+
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ListMultimap;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Optional;
+
+import static io.airlift.http.client.ResponseHandlerUtils.propagate;
+import static java.util.Objects.requireNonNull;
+
+public class StreamingResponseHandler
+        implements ResponseHandler<StreamingResponseHandler.StreamingResponse, RuntimeException>
+{
+    private volatile Runnable completer;
+
+    public static StreamingResponseHandler streamingResponseHandler()
+    {
+        return new StreamingResponseHandler();
+    }
+
+    private StreamingResponseHandler() {}
+
+    public final class StreamingResponse
+            implements Closeable
+    {
+        private final int statusCode;
+        private final ListMultimap<HeaderName, String> headers;
+        private final Response response;
+
+        private StreamingResponse(Response response)
+        {
+            this.response = requireNonNull(response, "response is null");
+            this.statusCode = response.getStatusCode();
```

Comment:
No need to copy these fields since we're holding onto `Response`. We can say that these are expected to be immutable in `Response` as we own the implementation.

Or, we could also fetch `InputStream` in the constructor and not hold `Response`, since there is no need for that call to be lazy. Either way is fine with me.

Reply:
I wasn't sure the consequences of accessing the `InputStream` early. Let's get rid of the fields and just use `response`.